docs: Update the documentation section for the IntelliJ-based IDEs (#2104)
codenvy/che-docs: modules/end-user-guide/partials/proc_building-an-image-for-an-ide-based-on-the-intellij-platform.adoc

[id="building-an-image-for-an-ide-based-on-the-intellij-platform_{context}"]
= Building an image for an IDE based on the IntelliJ Platform

> This procedure describes how to build an image for an IDE based on the IntelliJ Platform up to version 2020.3. See the link:https://github.com/che-incubator/jetbrains-editor-images/blob/main/doc/Compatible-IDE.md[list of compatible IDEs].

Reviewer: The current line reads "_based on the **IntelliJ Platform up to version 2020.3**_".

Reviewer: ```suggestion
This procedure describes how to build an image for a JetBrains IDE up to version 2020.3. See the link:https://github.com/che-incubator/jetbrains-editor-images/blob/main/doc/Compatible-IDE.md[list of compatible IDEs].
```

Reviewer 2: Why remove the link to the list of compatible IDEs?

Reviewer: @themr0c Sorry about that, a copy-paste gone wrong on the suggestion.

Reviewer 2: Discarding this suggestion as we have clearly established that  JetBrains IDE s are only a subset of all IDEs based on the IntelliJ Platform that are supported (Android Studio).

Reviewer: @themr0c I believe it's not possible to mention _**IDE based on the IntelliJ Platform up to version 2020.3**_ for 2 reasons:
1. Only JetBrains IDEs have versions 2020.3. Non-JetBrains IDEs that are based on the IntelliJ Platform most likely have other versioning.
2. **_version 2020.3_** is the version of the IDEs, not of the IntelliJ Platform.
My suggestion is now corrected for these reasons.

Reviewer: @themr0c In other words, the way I understand this, where you mention 2020.3 you are referring to JetBrains IDEs.

.Prerequisites

* The build host has at least 2 GB of available RAM.

> * The build host has a local installation of following tools:

Reviewer: `installation` = the act of putting a program onto a computer (source: dictionary.cambridge.org)

imho, `local` (as in *locally installed*) is redundant and can be dropped if the location/host is specified (here `the build host`).

Reviewer 2: fixed.

> * The build host has an installation of following tools:

Reviewer: `The build host has an installation of following tools:` = `The build host makes/undergoes an installation of the following tools:`

Reviewer: ```suggestion
* The following tools are installed on the build host:
```

Reviewer: https://redhat-documentation.github.io/supplementary-style-guide/#general-formatting
_"When writing prerequisites, be as clear and concise as possible. **You can use the passive voice, if necessary, to achieve that end.**"_

Reviewer 2: Oh that's good to know.

** link:https://www.docker.com/[Docker] version 18.09 or greater, supporting link:https://docs.docker.com/develop/develop-images/build_enhancements/[Docker BuildKit]

** link:https://git-scm.com[Git]

** link:https://www.gnu.org/software/libc/manual/html_node/Getopt.html[GNU getopt]

** link:https://www.gnu.org/software/wget/[GNU wget]

** link:https://openjdk.java.net/projects/jdk/11/[Java Development Kit (JDK) version 11]

** link:https://stedolan.github.io/jq/[jq]

.Procedure

. Get a local copy of the link:https://github.com/che-incubator/jetbrains-editor-images[JetBrains Projector Editor Images repository].
+
----
$ git clone https://github.com/che-incubator/jetbrains-editor-images
$ cd jetbrains-editor-images
----

. Run the build script with parameters:
+
[subs="+quotes,macros,attributes"]
----
$ ./projector build --tag __<tag>__ --url __<URL>__
----
+
`--tag __<tag>__`::
The name and tag to apply to the image after build in `name:tag` format.
+
`--url __<URL>__`::

> The URL pointing to an archive of the custom IntelliJ-based distribution. The editor version can be at most 2020.3 because Projector doesn't support later versions. The archive must target the Linux platform, be in `+tar.gz+` format, and include JetBrains Runtime (JBR). 

Reviewer: ```suggestion
The URL pointing to an archive of the custom IntelliJ Platform-based IDE distribution. The IDE version can be at most 2020.3 because Projector doesn't support later versions. The archive must be a `+tar.gz+`, target the Linux platform, and include JetBrains Runtime (JBR). 
```

Reviewer 2: A Platform-based contraction that I forgot to replace.

The order matters: 1.  platform, 2. format, 3. content. It's the natural order when searching for an archive. When searching in JetBrains of Android Studio download website, you will find the information in that order. 

 "a  tar.gz" looks awkward.  `tar.gz` is really a file format.

Reviewer 3: > "a tar.gz" looks awkward. tar.gz is really a file format.

+1

+
.Building the Docker image with IntelliJ IDEA Community 2020.3.3
====
----
./projector.sh build --tag che-idea:2020.3.3 \
  --url https://download-cdn.jetbrains.com/idea/ideaIC-2020.3.3.tar.gz
----
====
+
.Building the Docker image with PyCharm Community 2020.3.5
====
----
$ ./projector.sh build --tag che-pycharm:2020.3.5 \
  --url https://download.jetbrains.com/python/pycharm-community-2020.3.5.tar.gz
----
====
+
.Building the Docker image with WebStorm 2020.3.3
====
----
$ ./projector.sh build --tag che-webstorm:2020.3.3 \
  --url https://download.jetbrains.com/webstorm/WebStorm-2020.3.3.tar.gz
----
====
+
.Building the Docker image with IntelliJ IDEA Ultimate 2020.2.2
====
----
$ ./projector.sh build --tag che-idea-ultimate:2020.2.2 \
  --url https://download.jetbrains.com/idea/ideaIU-2020.2.2.tar.gz
----
====

> . To test the image, run it locally, and navigate to `++http://localhost:8887++` to access the IDE.

Reviewer: IMB Style, p. 314:
_navigate verb
Use only when a more direct verb, such as "click", "select", "browse", or "go to", is not applicable. Use only to provide direction to a location (for example, "Navigate to the directory that contains the files that you want to edit")._
imho `go to` is easier to understand if you just mean opening a website.

Reviewer: ```suggestion
. To test the image, run it locally and go to `++http://localhost:8887++` to access the IDE.
```

+
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh run __<tag>__
----
+
.Testing the Docker image with IntelliJ IDEA Community 2020.3.3
====
----
$ ./projector.sh run che-idea:2020.3.3
----
====
+
.Testing the Docker image with PyCharm 2020.3.5
====
----
$ ./projector.sh run che-pycharm:2020.3.5
----
====
+
.Testing the Docker image with WebStorm 2020.3.3
====
----
$ ./projector.sh run che-webstorm:2020.3.3
----
====
+
.Testing the Docker image with IntelliJ IDEA Ultimate 2020.2.2
====
----
$ ./projector.sh run che-idea-ultimate:2020.2.2
----
====

> . Publish the image to a registry accessible by {prod-short}, and remember the location: __<image>__.

Reviewer: ```suggestion
. Publish the image to a registry accessible by {prod-short} and copy the image URL in the registry for use as `__<full_path_to_image>__` in the next step.
```

. Create a `+meta.yaml+` file with the following content:
+
[source,yaml,subs="+quotes,macros,attributes"]
----
apiVersion: v2
publisher: __<publisher>__ <1>
name: intellij-ide
version: latest
type: Che Editor
displayName: IntelliJ Platform IDE

> title: IDE based on the IntelliJ Platform as Editor for Eclipse Che

Reviewer: This is not clear to me at all. (`Editor` is already mentioned in `type` and `category`.)

description: IDE based on the IntelliJ Platform running using Projector
icon: https://www.jetbrains.com/apple-touch-icon.png
category: Editor
repository: https://github.com/che-incubator/jetbrains-editor-images
firstPublicationDate: "2021-04-10"
spec:
  endpoints:
   -  name: "intellij"
      public: true
      targetPort: 8887
      attributes:
        protocol: http
        type: ide
        path: /projector-client/index.html?backgroundColor=434343&wss
  containers:
   - name: intellij-ide

>      image: "__<image>__" <2>

Reviewer: ```suggestion
     image: "__<full_path_to_image>__" <2>
```

Reviewer: Double quotes?

Reviewer 2: YAML syntax allows no quotes, double quotes or single quotes.

     mountSources: true
     volumes:
         - mountPath: "/home/projector-user"
           name: projector-user
     ports:
         - exposedPort: 8887
     memoryLimit: "4096M"
----

> <1> __<publisher>__: Your publisher name.

Reviewer: ```suggestion
<1> `__<publisher>__`: Your publisher name.
```

> <2> __<image>__: Location of the editor image in a registry accessible by {prod-short}.

Reviewer: ```suggestion
<2> `__<image>__`: Location of the IDE image in a registry accessible by {prod-short}.
```

> . Publish the `meta.yaml` file to an HTTPS resource accessible by {prod-short}: `__<URL_to_meta.yaml>__`.

Reviewer: Looks like `__<URL_to_meta.yaml>__` is an orphan here.

Reviewer: ```suggestion
. Publish the `meta.yaml` file to an HTTPS resource accessible by {prod-short} and copy the full path to the file for use as `__<URL_to_meta.yaml>__` when configuring a workspace devfile for this IDE image.
```

Reviewer 2: `full path` is misleading.

Reviewer 3: No reason to instruct the user to 'copy'.

.Building the image with IntelliJ IDEA Ultimate 2020.2.2
====
----
$ ./projector.sh build --tag che-idea-ultimate:2020.2.2 \
  --url https://download.jetbrains.com/idea/ideaIU-2020.2.2.tar.gz
----
====
+
.Building the image with Android Studio 4.2.0.22
====
----
$ ./projector.sh build --tag che-android-studio:4.2.0.22 \
  --url https://redirector.gvt1.com/edgedl/android/studio/ide-zips/4.2.0.22/android-studio-ide-202.7188722-linux.tar.gz
----
====

. To test the image, run it locally and go to `++http://localhost:8887++` to access the IDE.
+
[subs="+quotes,macros,attributes"]
----
$ ./projector.sh run __<tag>__
----
+
.Testing the image with IntelliJ IDEA Community 2020.3.3
====
----
$ ./projector.sh run che-idea:2020.3.3
----
====
+
.Testing the image with PyCharm 2020.3.5
====
----
$ ./projector.sh run che-pycharm:2020.3.5
----
====
+
.Testing the image with WebStorm 2020.3.3
====
----
$ ./projector.sh run che-webstorm:2020.3.3
----
====
+
.Testing the image with IntelliJ IDEA Ultimate 2020.2.2
====
----
$ ./projector.sh run che-idea-ultimate:2020.2.2
----
====
+
.Testing the image with Android Studio 4.2.0.22
====
----
$ ./projector.sh run che-android-studio:4.2.0.22
----
====

> . Publish the image to a registry accessible by {prod-short}, and remember the location: __<registry>/<image>:<tag>__.

Reviewer: 1. `remember the location` for when/which step?
2.  Suggestion instead of `remember the location` > `copy the resulting URL for use (as/when ...)`

. Create a `+meta.yaml+` file containing the IDE metadata for {prod-short}:
+
[source,yaml,subs="+quotes,macros,attributes"]
----
apiVersion: v2
publisher: __<publisher>__ <1>
name: intellij-ide
version: latest
type: Che Editor
displayName: IntelliJ Platform IDE
title: IDE based on the IntelliJ Platform
description: IDE based on the IntelliJ Platform running using Projector
icon: https://www.jetbrains.com/apple-touch-icon.png
category: Editor
repository: https://github.com/che-incubator/jetbrains-editor-images
firstPublicationDate: "2021-04-10"
spec:
  endpoints:
   -  name: intellij
      public: true
      targetPort: 8887
      attributes:
        protocol: http
        type: ide
        path: /projector-client/index.html?backgroundColor=434343&wss
  containers:
   - name: intellij-ide
     image: __<registry>/<image>:<tag>__ <2>
     mountSources: true
     volumes:
         - mountPath: "/home/projector-user"
           name: projector-user
     ports:
         - exposedPort: 8887
     memoryLimit: "4096M"
----
<1> `__<publisher>__`: Your publisher name.
<2> `__<registry>/<image>:<tag>__`: Location of the IDE image in a registry accessible by {prod-short}.

> . Publish the `meta.yaml` file to an HTTPS resource accessible by {prod-short} and remember the resulting URL as `__<URL_to_meta.yaml>__` when configuring a workspace to use this IDE.

Reviewer: ```suggestion
. Publish the `meta.yaml` file to an HTTPS resource accessible by {prod-short} and copy the resulting URL for use as `__<URL_to_meta.yaml>__` when configuring a workspace to use this IDE.
```
